Feature - custom response handler support

andyskw opened the pull request:
I've introduced a new option to support custom responses, as with the current implementation it seemed impossible to handle SAMLResponse via XHR (or at least without postprocessing the HTML output).

The use case I was trying to cover with the change was that if you are trying to achieve a Zero Page Login flow with SAML2  via XHR requests. 
A good example for this is the following scenario: log in a user in a remote SP based on a current session token, but you are not aware of the user's credentials in the remote service. In this case you can POST the SAMLResponse via XHR, and extracting it from a HTML output makes no sense. Instead you could pass it around as JSON or an XML, or whatever format you wish, and the original caller can POST it to the original returnURL.

Please check out my changes, and if you consider them safe (not violating any protocol constraints), and it seems a reasonable change, consider merging it to the main library.

siacomuzzi commented on README.md:
````diff
@@ -31,6 +31,7 @@ Options
 | digestAlgorithm     | digest algorithm, options: sha1, sha256          | ```'sha256'``` |
 | RelayState          | state of the auth process                        | ```req.query.RelayState || req.body.RelayState``` |
 | sessionIndex          | the index of a particular session between the principal identified by the subject and the authenticating authority                        | _SessionIndex is not included_ |
+| responseSender        | custom response sender; signature: fn(req,res,samlresponse)| HTML response that POSTS to returnUrl |
````
use `getPostUrl` syntax. Something like:

`| responseSender | custom SAML response sender; f(req, res, SAMLResponse) | HTML response that POSTS to postUrl |`

siacomuzzi commented:
looks good to me. some minor feedback:

* we have the `getPostUrl` method, so I think this new one should be named in a similar way. Maybe `post`?
* please udpate `auth` method documentation: https://github.com/andyskw/node-samlp/blob/feature/responseSender/lib/samlp.js#L186-L192

andyskw commented:
@siacomuzzi Thanks for the feedback!

I've updated the documentation, and fixed the missing doc as well in samlp.js.

About the function name - I am not sure; however I don't really like the ```responseSender``` either by the way, but I feel like ```post``` is not a valid name, since the plugin allows you other ways to take care of the SAMLResponse.
For example, a 

    res.json({SAMLResponse: samlResp.toString("base64")});

is possible via this change (obviously the receiver must POST the received data to the IdP, but it allows you  to pull this responsibility out from the library, if you wish so).

On the other hand, if you still think, that ```post``` is a good choice, then I am open for changing it.
Although I'd maybe use ```responseHandler``` instead, or ```customResponseHandler```, or something similar.

What do you think?

siacomuzzi commented:
`responseHandler` :+1:

andyskw commented:
Yip, I like that one as well, probably that's what we are looking for. :)

PR updated!

andyskw commented:
Yaay, thanks! :)

siacomuzzi commented:
available @ `samlp@0.12.3`

thanks for contributing!
